lib/vdc/DerivedParticleDensity.cpp

```diff
@@ -0,0 +1,355 @@
+#include <vapor/DerivedParticleDensity.h>
+#include <vapor/STLUtils.h>
+#include <vapor/DataMgrUtils.h>
+#include <float.h>
+#include <glm/glm.hpp>
+
+using namespace VAPoR;
+using glm::ivec3;
+using glm::vec3;
+
+// ===================================
+//       DerivedParticleDensity
+// ===================================
+
+DerivedParticleDensity::DerivedParticleDensity(string varName, DC *dc, string meshName, DataMgr *dataMgr) : DerivedDataVar(varName), _dc(dc), _meshName(meshName), _dataMgr(dataMgr) {}
+
+int DerivedParticleDensity::Initialize()
+{
+    auto   dataVarNames = _dataMgr->GetDataVarNames(3);
+    string dataVar;
+    for (auto dv : dataVarNames) {
+        vector<string> coordNames;
+        _dataMgr->GetVarCoordVars(dv, true, coordNames);
+        if (STLUtils::Contains(coordNames, string("Position_x")) && STLUtils::Contains(coordNames, string("Position_y")) && STLUtils::Contains(coordNames, string("Position_z"))) {
+            dataVar = dv;
+            break;
+        }
+    }
+
+    if (dataVar.empty()) {
+        assert(0);
+        return -1;
+    }
+
+    _dataVar = dataVar;
+
+    return 0;
+}
+
+int DerivedParticleDensity::OpenVariableRead(size_t ts, int level, int lod)
+{
+    DC::FileTable::FileObject *f = new DC::FileTable::FileObject(ts, _derivedVarName, level, lod);
+    return _fileTable.AddEntry(f);
+}
+
+int DerivedParticleDensity::CloseVariable(int fd)
+{
+    DC::FileTable::FileObject *f = _fileTable.GetEntry(fd);
+    if (!f) {
+        SetErrMsg("Invalid file descriptor : %d", fd);
+        return -1;
+    }
+    _fileTable.RemoveEntry(fd);
+    delete f;
+    return 0;
+}
+
+int DerivedParticleDensity::ReadRegion(int fd, const std::vector<size_t> &min, const std::vector<size_t> &max, float *region)
+{
+    DC::FileTable::FileObject *f = _fileTable.GetEntry(fd);
+    if (!f) {
+        SetErrMsg("Invalid file descriptor : %d", fd);
+        return -1;
+    }
+
+
+    vector<size_t> dims;
+    _dataMgr->GetDimLens(_derivedVarName, dims, f->GetTS());
+    if (dims.size() != 3) {
+        assert(0);
+        return -1;
+    }
+
+    Grid *g = _dataMgr->GetVariable(f->GetTS(), _dataVar, -1, -1, true);
+    if (!g) {
+        assert(0);
+        return -1;
+    }
+
+    vector<size_t> particleDims;
+    _dataMgr->GetDimLens(_dataVar, particleDims, f->GetTS());
+    assert(particleDims.size() == 1);
+    size_t realNP = particleDims[0];
+
+    int    xd = dims[0], yd = dims[1], zd = dims[2];
+    float *data = new float[xd * yd * zd];
+    compute(g, data, xd, yd, zd, realNP);
+    delete g;
+
+    //    assert(max[0]-min[0]+1 == xd);
+    //    assert(max[1]-min[1]+1 == yd);
+    //    assert(max[2]-min[2]+1 == zd);
+
+    //    for (int z = 0; z < zd; z++)
```

Comment:
Can we remove this?

Reply:
This is part of an in-progress prototype so it has some commented out code that I was hoping to leave but I can remove it if you insist.